### ISSUE TITLE
Define GLSL variables

### ISSUE THREAD

**Issue by lysek** [openlayers/openlayers#14350]
**Is your feature request related to a problem? Please describe.**
I used the WebGL terrain shading from this [example](https://openlayers.org/en/latest/examples/webgl-shaded-relief.html) but on some older devices (Samsung Note 3) the shader compilation reported *Too many vertex shader constants* error during compilation. I ended with custom GLSL operators defined into `Operators` object from `ol/style/expressions` but my solution is bulky and messy, it combines custom GLSL function defined on `context` and generating code with `expressionToGlsl()` function. The whole thing would be easier if it was possible to define a variable which can be reused in shader.

The main problem is, that generated GLSL expression is 3x the same formula with a lot of constants due to this expression `color: ['color', scaled, scaled, scaled]` in the layer style definition, when I changed it to `color: ['color', scaled, 0, 0]` it worked on that old Samsung, but the shading was only in red color.

Anyway calculating the same thing three times is not very efficient.

**Describe the solution you'd like**
I propose the ability to define and use float variable and reuse it in shaders, e.g.:

``` javascript
//code from terrain shading example

const cosIncidence = [
  '+',
  ['*', ['sin', sunEl], ['cos', slope]],
  ['*', ['*', ['cos', sunEl], ['sin', slope]], ['cos', ['-', sunAz, aspect]]],
];

const shadedRelief = new TileLayer({
  opacity: 0.3,
  source: new XYZ({
    url: 'https://{a-d}.tiles.mapbox.com/v3/aj.sf-dem/{z}/{x}/{y}.png',
  }),
  style: {
    //variables from outside (uniforms)
    variables: variables,
    //variables with intermediate calculations inside shader
    floats: {
      scaled: ['*', 255, cosIncidence]
    },
    //use the same float value
    color: ['color', ['varFloat', 'scaled'], ['varFloat', 'scaled'], ['varFloat', 'scaled']],
  },
});
```

It should be easy to generate few lines for the shader code in the `WebGLTile.js` file inside the `parseStyle()` function to define `float prefix_varName = ${expressionToGlsl(...)};` just before `${pipeline.join('\n')}` in the fragment shader and the operator `['varFloat', 'name']` would just output the `prefix_varName` to where it was used.

**Comment by jahow** [comment]
This is definitely a shortcoming of the GLSL shader generator. I feel we could also address this more widely by looking for identical parts in the expression tree, and factorizing them in the GLSL code. Then you could just do:

```js
const cosIncidence = [
  '+',
  ['*', ['sin', sunEl], ['cos', slope]],
  ['*', ['*', ['cos', sunEl], ['sin', slope]], ['cos', ['-', sunAz, aspect]]],
];

const scaledIncidence =  ['*', 255, cosIncidence];

const shadedRelief = new TileLayer({
  opacity: 0.3,
  source: new XYZ({
    url: 'https://{a-d}.tiles.mapbox.com/v3/aj.sf-dem/{z}/{x}/{y}.png',
  }),
  style: {
    variables: variables,
    color: ['color', scaledIncidence, scaledIncidence, scaledIncidence],
  },
});
```

Without worrying about implementation details (such as intermediary variables). This would also have an effect on all existing code.

Would that work for you?

**Comment by jahow** [comment]
> Given the placeholder operator would have to be created anyway, it would be nice to let others to also define their own variables as I proposed.
> 
> I think that first step is to let users optimize the expression by themselves and second step can be to automate it somehow.

Hmm, I'm still not sure an "intermediary variable" operator would make sense here. Expressions are supposed to be declarative and not contain any programming-like considerations, but intermediary variables really are just that, they do not produce anything meaningful in the final result and are just an indication on how to get there more efficiently.

I will pick up work on webgl renderer very soon so this could be an interesting performance improvement to tackle as part of it. Thanks for bringing this topic up!

**Comment by jahow** [comment]
Thanks for your feedback. Would an expression like this work for you?

```js
['*',
  ['color', cosIncidence, cosIncidence, cosIncidence, 1],
  ['color', interpolatedR, interpolatedG, interpolatedB, 1]
]
```

(note the final `1` is optional, as it is the default alpha value)

As for what  you proposed:
* identifying identical subtrees is still something I'd like to do at some point, as that could make resulting shaders much simpler
* defining custom GLSL operators doesn't sound like the way to got for me; instead I would rather open up usage of the `ShaderBuilder` class (even though it might not be part of the official API) to give people more control over the final shaders
* adding a `mulcolor` operator seems like not a great response to the problem; instead being able to create a color with just one scalar might be more handy? e.g. like `['color', cosIncidence]`; then the regular `*` operator would do the job
* same for the `alpha` operator; it sounds like we could achieve everything we want with just the `color` and `*` operators

**Comment by lysek** [comment]
The problem with `['color', cosIncidence, cosIncidence, cosIncidence, 1]` is that it generates three times the same code and it fails on legacy devices with *Too many vertex shader constants* error message.

The possibility to supply just one argument to `color` operator seems OK if that means the subtree is evaluated only once. There should also be the options to specify alpha:
- `['color', shade]`
- `['color', shade, alpha]`
- `['color', r, g, b]`
- `['color', r, g, b, alpha]`

This could result in following style which I expect to work for me:
```
['*',
  ['color', cosIncidence, 1],
  ['color', interpolatedR, interpolatedG, interpolatedB, 1]
]
```

**Comment by jahow** [comment]
Yes exactly, when providing only one scalar to `color` the value would only be evaluated once and then used for all color channels. Would you be ready to try your hand at a PR doing that?

**Comment by lysek** [comment]
OK, I will look at the code tomorrow and try to extend the `color` operator. It seems like a small change.

**Comment by jahow** [comment]
Closed by #15359

**Comment by lysek** [comment]
I will just drop my shading with heatmap style for WebGL tile-layer as an example for others, this is optimized to generate short fragment shader code and runs even on legacy device with poor WebGL specs:

```javascript
const variables = {
    vert: 5,
    sunEl: Math.PI / 4,
    sunAz: -Math.PI / 2
};

//elevation data is encoded in 2 band array as unsigned integer and increased by 10000 meters to be > 0
function elevation(xOffset = 0, yOffset = 0) {
    const a = ['band', 1, xOffset, yOffset];
    const b = ['band', 2, xOffset, yOffset];
    return ['+', ['*', 65536, a], ['*', 256, b]];
}

//palette for altitudes from 0 to 3600 meters
const interpolateRGB = [
    'interpolate',
    ['linear'],
    elevation(),
    10000, ["color",0,0,255],    //0 meters above sea level
    10600, ["color",0,255,255],
    11200, ["color",0,255,0],
    11800, ["color",255,255,0],
    12400, ["color",255,0,0],
    13000, ["color",255,255,255],
    13600, ["color",255,0,255]  //3600 meters above sea level
];

//for shading, the 10000 meters offset is irrelevant, it requires only the difference
const dzdx = ['/', ['*', ['var', 'vert'], ['-', elevation(1, 0), elevation(-1, 0)], 0.5], ['resolution']];
const dzdy = ['/', ['*', ['var', 'vert'], ['-', elevation(0, 1), elevation(0, -1)], 0.5], ['resolution']];
const slope = ['atan', ['sqrt', ['+', ['^', dzdx, 2], ['^', dzdy, 2]]]];
const aspect = ['clamp', ['atan', ['-', 0, dzdx], dzdy], -Math.PI, Math.PI];
const cosIncidence = [
    '+',
    ['*', 255, ['sin', ['var', 'sunEl']], ['cos', slope]],
    ['*', 255, ['cos', ['var', 'sunEl']], ['sin', slope], ['cos', ['-', ['var', 'sunAz'], aspect]]]
];

const styleShade = ['color', cosIncidence];
const styleHeatmap = interpolateRGB;
const styleCombined = ['*', interpolateRGB, ['color', cosIncidence]];
```

@jahow I still see that there is a space to improve the expression parser; for example the expression `styleCombined` generates following code - watch for the calls of `getBandValue` function and the repetetive decoding of altitude:

```glsl
vec4 color = texture2D(u_tileTextures[0],  v_textureCoord);
color = (mix(mix(mix(mix(mix(mix(vec4(0.0 / 255.0, 0.0 / 255.0, 255.0 / 255.0, 1.0), vec4(0.0 / 255.0, 255.0 / 255.0, 255.0 / 255.0, 1.0), clamp((((65536.0 * getBandValue(1.0, 0.0, 0.0)) + (256.0 * getBandValue(2.0, 0.0, 0.0))) - 10000.0) / (10600.0 - 10000.0), 0.0, 1.0)), vec4(0.0 / 255.0, 255.0 / 255.0, 0.0 / 255.0, 1.0), clamp((((65536.0 * getBandValue(1.0, 0.0, 0.0)) + (256.0 * getBandValue(2.0, 0.0, 0.0))) - 10600.0) / (11200.0 - 10600.0), 0.0, 1.0)), vec4(255.0 / 255.0, 255.0 / 255.0, 0.0 / 255.0, 1.0), clamp((((65536.0 * getBandValue(1.0, 0.0, 0.0)) + (256.0 * getBandValue(2.0, 0.0, 0.0))) - 11200.0) / (11800.0 - 11200.0), 0.0, 1.0)), vec4(255.0 / 255.0, 0.0 / 255.0, 0.0 / 255.0, 1.0), clamp((((65536.0 * getBandValue(1.0, 0.0, 0.0)) + (256.0 * getBandValue(2.0, 0.0, 0.0))) - 11800.0) / (12400.0 - 11800.0), 0.0, 1.0)), vec4(255.0 / 255.0, 255.0 / 255.0, 255.0 / 255.0, 1.0), clamp((((65536.0 * getBandValue(1.0, 0.0, 0.0)) + (256.0 * getBandValue(2.0, 0.0, 0.0))) - 12400.0) / (13000.0 - 12400.0), 0.0, 1.0)), vec4(255.0 / 255.0, 0.0 / 255.0, 255.0 / 255.0, 1.0), clamp((((65536.0 * getBandValue(1.0, 0.0, 0.0)) + (256.0 * getBandValue(2.0, 0.0, 0.0))) - 13000.0) / (13600.0 - 13000.0), 0.0, 1.0)) * vec4(vec3(((255.0 * sin(u_var_sunEl) * cos(atan(sqrt((pow(((u_var_vert * (((65536.0 * getBandValue(1.0, 1.0, 0.0)) + (256.0 * getBandValue(2.0, 1.0, 0.0))) - ((65536.0 * getBandValue(1.0, -1.0, 0.0)) + (256.0 * getBandValue(2.0, -1.0, 0.0)))) * 0.5) / u_resolution), 2.0) + pow(((u_var_vert * (((65536.0 * getBandValue(1.0, 0.0, 1.0)) + (256.0 * getBandValue(2.0, 0.0, 1.0))) - ((65536.0 * getBandValue(1.0, 0.0, -1.0)) + (256.0 * getBandValue(2.0, 0.0, -1.0)))) * 0.5) / u_resolution), 2.0)))))) + (255.0 * cos(u_var_sunEl) * sin(atan(sqrt((pow(((u_var_vert * (((65536.0 * getBandValue(1.0, 1.0, 0.0)) + (256.0 * getBandValue(2.0, 1.0, 0.0))) - ((65536.0 * getBandValue(1.0, -1.0, 0.0)) + (256.0 * getBandValue(2.0, -1.0, 0.0)))) * 0.5) / u_resolution), 2.0) + pow(((u_var_vert * (((65536.0 * getBandValue(1.0, 0.0, 1.0)) + (256.0 * getBandValue(2.0, 0.0, 1.0))) - ((65536.0 * getBandValue(1.0, 0.0, -1.0)) + (256.0 * getBandValue(2.0, 0.0, -1.0)))) * 0.5) / u_resolution), 2.0))))) * cos((u_var_sunAz - clamp(atan((0.0 - ((u_var_vert * (((65536.0 * getBandValue(1.0, 1.0, 0.0)) + (256.0 * getBandValue(2.0, 1.0, 0.0))) - ((65536.0 * getBandValue(1.0, -1.0, 0.0)) + (256.0 * getBandValue(2.0, -1.0, 0.0)))) * 0.5) / u_resolution)), ((u_var_vert * (((65536.0 * getBandValue(1.0, 0.0, 1.0)) + (256.0 * getBandValue(2.0, 0.0, 1.0))) - ((65536.0 * getBandValue(1.0, 0.0, -1.0)) + (256.0 * getBandValue(2.0, 0.0, -1.0)))) * 0.5) / u_resolution)), -3.141592653589793, 3.141592653589793))))) / 255.0), 1.0));
```

Thank you.
